File: src/main/java/org/cip4/xjdf/json/openapi/SimpleType.java
package org.cip4.xjdf.json.openapi;

import lombok.RequiredArgsConstructor;
import org.cip4.xjdf.json.openapi.model.Model;
import org.jetbrains.annotations.NotNull;
import org.w3c.dom.Node;

import javax.xml.xpath.XPath;
import java.util.List;
import java.util.UUID;

@RequiredArgsConstructor
public abstract class SimpleType implements Modelable {
    @NotNull
    protected Node node;

    public String getName() {
        return node.getAttributes().getNamedItem("name") != null
            ? node.getAttributes().getNamedItem("name").getNodeValue()
            : "S" + UUID.randomUUID();
    }

    @Override
    public abstract Model getModel();

    public interface Factory<T extends SimpleType> {
        boolean supports(Node candidate, XPath xPath);

        T create(Node node, Context context);
    }

    static SimpleType create(Node node, Context context) {
        List<Factory<? extends SimpleType>> factories = List.of(
            new NamedColor.Factory(),
            new SimpleTypeRestriction.Factory(),
            new SimpleTypeList.Factory(),
            new SimpleTypeLocalList.Factory()
        );
        for (Factory<? extends SimpleType> factory : factories) {
            if (factory.supports(node, context.getXPath())) {
                return factory.create(node, context);
            }
        }
        throw new IllegalStateException("Unsupported attribute detected...");
    }
}

File: src/main/kotlin/org/cip4/xjdf/openapi/SimpleType.kt
package org.cip4.xjdf.openapi

import org.cip4.xjdf.openapi.model.Schema
import org.w3c.dom.Node
import javax.xml.xpath.XPath

abstract class SimpleType(
    val node: Node
) {

    val name: String?
        get() = node.attributes.getNamedItem("name")?.nodeValue

    abstract fun getModel(): Schema

    interface Factory<T : SimpleType> {
        fun supports(candidate: Node, xPath: XPath): Boolean

        fun create(node: Node, context: Context): T

        companion object {
            fun create(node: Node, context: Context): SimpleType {
                val factories = listOf(
                    SimpleTypeRestriction.Factory,
                    SimpleTypeList.Factory,
                    SimpleTypeLocalList.Factory
                )
                val factory = factories.find { factory ->
                    factory.supports(node, context.xPath)
                } ?: throw IllegalStateException("Unsupported attribute detected...")
                return factory.create(node, context)
            }
        }
    }
}
